dotCMS/src/main/java/com/dotmarketing/portlets/folders/business/FolderAPIImpl.java

```diff
@@ -50,826 +51,839 @@
+            final String path = folder.getPath() != null ? folder.getPath() : id.getParentPath();
+            Logger.error(this, "User " + userId  + " does not have permission to read Folder " + path + " Please check the folder exists.");
+            throw new DotSecurityException("User " + userId + " does not have permission to read Folder " + path);
+        }
+        return folder;
+
+    }
+
+    /**
+     *
+     * @param folder
+     * @return List of sub folders for passed in folder
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    @CloseDBIfOpened
+    public List<Folder> findSubFolders(final Folder folder, final User user,
+                                       final boolean respectFrontEndPermissions) throws DotDataException,
+            DotSecurityException {
+
+        if (!doesUserHavePermissions(folder, PermissionAPI.PERMISSION_READ, user, respectFrontEndPermissions)) {
+
+            throw new DotSecurityException("User " + user.getUserId() != null?user.getUserId():"" + " does not have permission to read Folder " + folder.getPath());
+        }
+
+        return this.folderFactory.getFoldersByParent(folder, user, respectFrontEndPermissions);
+    } // findSubFolders.
+
+    /**
+     *
+     * @param folder
+     * @return List of sub folders for passed in folder
+     * @throws DotSecurityException
+     * @throws DotDataException
+     */
+    @CloseDBIfOpened
+    public List<Folder> findSubFolders(final Host host, final User user,
+                                       final boolean respectFrontEndPermissions) throws DotDataException,
+            DotSecurityException {
+
+        if (!this.permissionAPI.doesUserHavePermission(host, PermissionAPI.PERMISSION_READ, user, respectFrontEndPermissions)) {
+            throw new DotSecurityException("User " + user.getUserId() != null?user.getUserId():"" + " does not have permission to read Host " + host.getHostname());
+        }
+
+        final List<Folder> allFolders = folderFactory.findFoldersByHost(host);
+        
+        return allFolders.stream().filter (folder ->  
+            doesUserHavePermissions(folder, PermissionAPI.PERMISSION_READ, user, respectFrontEndPermissions)
+                ).collect(CollectionsUtils.toImmutableList()); 
+        
+    }
+    
+    @CloseDBIfOpened
+    public List<Folder> findThemes(final Host host, final User user,
+                                   final boolean respectFrontEndPermissions) throws DotDataException,
+    DotSecurityException {
+
+        if (!this.permissionAPI.doesUserHavePermission(host, PermissionAPI.PERMISSION_READ, user, respectFrontEndPermissions)) {
+            throw new DotSecurityException("User " + user.getUserId() != null?user.getUserId():"" + " does not have permission to read Host " + host.getHostname());
+        }
+
+        List<Folder> themesByHost = folderFactory.findThemesByHost(host);
+        
+        return themesByHost.stream().filter (folder ->  
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Specify a type for: 'folder' [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2211)